include/cetl/span.h

```diff
@@ -1 +1,772 @@
-// TODO: span for C++14.
\ No newline at end of file
+/// @file
+/// Defines a generic span type
+/// @copyright Copyright (c) 2023 Amazon.com Inc. and its affiliates. All Rights Reserved.
+
+#ifndef CETL_SPAN_H_INCLUDED
+#define CETL_SPAN_H_INCLUDED
+
+#include <array>
+#include <iterator>
+#include <limits>
+#include <type_traits>
+
+#include "cetl/cetl.h"
+
+namespace cetl
+{
+
+///
+/// Used by span to indicate that the span size is not fixed.
+/// @see std::dynamic_extent
+inline constexpr std::size_t dynamic_extent = std::numeric_limits<std::size_t>::max();
+
+/// A borrowed view into a contiguous set of objects. Spans can either be static, where the set of objects is fixed
+/// and known, or dynamic where the number of objects in the contiguous set may change.
+/// This template is compatible with class: std::span available in C++20.
+template <typename T, std::size_t Extent = cetl::dynamic_extent>
+class span;
+
+///
+/// This version is for spans where the extent is static (see span< T, cetl::dynamic_extent > for the dynamic
+/// extent specialization)
+/// @tparam T       The element type.
+/// @tparam Extent  The extent type of this span; either dynamic or static.
+template <typename T, std::size_t Extent>
+class span
+{
+public:
+    // (Groupings per https://en.cppreference.com/w/cpp/container/span)
+    // +----------------------------------------------------------------------+
+    // | Member Constants
+    // +----------------------------------------------------------------------+
+    ///
+    /// The value of Extent for this template instantiation.
+    static constexpr std::size_t extent = Extent;
+
+    // +----------------------------------------------------------------------+
+    // | Member Types
+    // +----------------------------------------------------------------------+
+    using element_type     = T;                                 //!< The element type over which the span operates
+    using value_type       = typename std::remove_cv<T>::type;  //!< The non-cv type of elements
+    using pointer          = T*;                                //!< The element pointer type
+    using const_pointer    = const T*;                          //!< A constant pointer to the element type
+    using reference        = T&;                                //!< A reference type to the element
+    using const_reference  = const T&;                          //!< A constant reference type to the element
```

Comment:
```suggestion
    using reference_const  = T const&;                          //!< A reference to the constant element type
```

Reply:
same as before. I'll resolve this conversation and we'll talk about your proposal versus the C++ standard up there.